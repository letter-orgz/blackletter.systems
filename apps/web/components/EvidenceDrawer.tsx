"use client";
import { useState } from "react";

export default function EvidenceDrawer() {
  const [open, setOpen] = useState(false);
  return (
    <>
      <button className="rounded border px-3 py-1" onClick={() => setOpen(true)}>Open Evidence</button>
      {open && (
        <div className="fixed inset-0 bg-black/40">
          <div className="absolute right-0 top-0 h-full w-full max-w-xl bg-white p-6 shadow-2xl">
            <button className="text-sm underline" onClick={() => setOpen(false)}>Close</button>
            <h3 className="text-lg font-medium mt-4">Evidence</h3>
            <p className="text-sm text-neutral-600">Wire this to a selected finding later.</p>
          </div>
        </div>
      )}
    </>
  );
}
