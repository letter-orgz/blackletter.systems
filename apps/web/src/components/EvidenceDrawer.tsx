'use client';

import React, { useEffect, useRef, useState } from 'react';

interface EvidenceDrawerProps {
  isOpen: boolean;
  onClose: () => void;
  children: React.ReactNode;
}

export default function EvidenceDrawer({ isOpen, onClose, children }: EvidenceDrawerProps) {
  const drawerRef = useRef<HTMLDivElement>(null);
  const snippetRef = useRef<HTMLDivElement>(null);
  const [copied, setCopied] = useState(false);

  const handleCopy = async () => {
    const text = snippetRef.current?.textContent ?? '';
    try {
      await navigator.clipboard.writeText(text);
      setCopied(true);
      setTimeout(() => setCopied(false), 2000);
    } catch (e) {
      console.error('Copy failed', e);
    }
  };

  useEffect(() => {
    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        onClose();
      } else if (e.key === 'Tab' && drawerRef.current) {
        const focusable = drawerRef.current.querySelectorAll<HTMLElement>(
          'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
        );
        if (focusable.length === 0) return;
        const first = focusable[0];
        const last = focusable[focusable.length - 1];
        if (e.shiftKey) {
          if (document.activeElement === first) {
            e.preventDefault();
            last.focus();
          }
        } else {
          if (document.activeElement === last) {
            e.preventDefault();
            first.focus();
          }
        }
      }
    };

    if (isOpen) {
      drawerRef.current?.focus();
      document.addEventListener('keydown', handleKeyDown);
    }
    return () => document.removeEventListener('keydown', handleKeyDown);
  }, [isOpen, onClose]);

  if (!isOpen) return null;

  return (
    <div className="fixed inset-0 z-40">
      <div className="fixed inset-0 bg-black/50" onClick={onClose} />
      <div
        ref={drawerRef}
        tabIndex={-1}
        className="absolute right-0 top-0 h-full w-80 bg-white p-4 shadow-lg"
      >
        <div ref={snippetRef}>{children}</div>
        <div className="mt-4 flex justify-end space-x-2">
          <button onClick={handleCopy}>Copy</button>
          <button onClick={onClose}>Close</button>
        </div>
        {copied && (
          <div
            role="status"
            className="absolute bottom-4 right-4 rounded bg-gray-800 px-3 py-2 text-white"
          >
            Copied to clipboard
          </div>
        )}
      </div>
    </div>
  );
}
