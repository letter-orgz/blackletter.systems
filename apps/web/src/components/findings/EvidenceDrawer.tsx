import type { Finding } from "@/hooks/useFindings";

export default function EvidenceDrawer({ finding, onClose }: { finding: Finding | null; onClose: () => void }) {
  return (
    <div className={`fixed inset-0 z-50 ${finding ? "pointer-events-auto" : "pointer-events-none"}`}>
      <div className={`absolute inset-0 bg-black/30 ${finding ? "opacity-100" : "opacity-0"}`} onClick={onClose} />
      <aside
        className={`absolute right-0 top-0 h-full w-full max-w-xl bg-white shadow-xl p-4 transition-transform ${
          finding ? "translate-x-0" : "translate-x-full"
        }`}
        aria-hidden={!finding}
      >
        {finding && (
          <div className="space-y-3">
            <h2 className="text-lg font-semibold">
              {finding.rule_id} <span className="text-gray-500 text-sm">(p.{finding.page})</span>
            </h2>
            <p className="text-sm text-gray-800 whitespace-pre-wrap">{finding.snippet}</p>
            {finding.rationale && <p className="text-xs text-gray-500">Rationale: {finding.rationale}</p>}
            <div className="flex justify-end pt-2">
              <button className="text-sm underline" onClick={onClose}>
                Close
              </button>
            </div>
          </div>
        )}
      </aside>
    </div>
  );
}

