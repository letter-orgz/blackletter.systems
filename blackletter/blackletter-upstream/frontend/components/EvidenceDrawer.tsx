'use client';

import { useState, useRef, useEffect } from 'react';
import { Dialog, DialogPanel, DialogTitle } from '@headlessui/react';
import { Finding } from '@/lib/types';
import { highlightAnchors } from '@/lib/anchors';
import { copyToClipboard } from '@/lib/utils';

interface EvidenceDrawerProps {
  isOpen: boolean;
  onClose: () => void;
  finding: Finding;
}

export default function EvidenceDrawer({ isOpen, onClose, finding }: EvidenceDrawerProps) {
  const [isReviewed, setIsReviewed] = useState(false);
  const drawerRef = useRef<HTMLDivElement>(null);

  // Handle ESC key to close the drawer
  useEffect(() => {
    const handleEsc = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        onClose();
      }
    };

    if (isOpen) {
      document.addEventListener('keydown', handleEsc);
      // Focus the drawer for accessibility
      drawerRef.current?.focus();
    }

    return () => {
      document.removeEventListener('keydown', handleEsc);
    };
  }, [isOpen, onClose]);

  // Highlight anchors in the evidence text
  const highlightedEvidence = highlightAnchors(finding.evidence, finding.anchors);

  const handleCopyToClipboard = async () => {
    await copyToClipboard(highlightedEvidence);
    // Could add a toast notification here
  };

  const handleMarkReviewed = () => {
    setIsReviewed(!isReviewed);
    // In a real app, this would update the finding's status in the store
  };

  return (
    <Dialog open={isOpen} onClose={onClose} className="relative z-50">
      {/* Backdrop */}
      <div className="fixed inset-0 bg-black/30" aria-hidden="true" />

      {/* Drawer panel */}
      <div className="fixed inset-0 overflow-hidden">
        <div className="absolute inset-0 overflow-hidden">
          <div className="pointer-events-none fixed inset-y-0 right-0 flex max-w-full pl-10">
            <DialogPanel
              ref={drawerRef}
              tabIndex={-1}
              className="pointer-events-auto w-screen max-w-md transform transition duration-300 ease-in-out data-[closed]:translate-x-full"
            >
              <div className="flex h-full flex-col divide-y divide-gray-200 bg-white shadow-xl">
                <div className="flex min-h-0 flex-1 flex-col overflow-y-scroll">
                  <div className="px-4 py-6 sm:px-6">
                    <div className="flex items-start justify-between">
                      <DialogTitle className="text-lg font-semibold text-gray-900">
                        Evidence for {finding.detector}
                      </DialogTitle>
                      <div className="ml-3 flex h-7 items-center">
                        <button
                          type="button"
                          className="rounded-md bg-white text-gray-400 hover:text-gray-500 focus:outline-none"
                          onClick={onClose}
                        >
                          <span className="sr-only">Close panel</span>
                          <svg className="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M6 18L18 6M6 6l12 12" />
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                  <div className="relative flex-1 px-4 py-6 sm:px-6">
                    {/* Evidence content */}
                    <div className="mb-6">
                      <h3 className="text-md font-medium text-gray-900 mb-2">Evidence</h3>
                      <div 
                        className="border p-4 rounded bg-gray-50 max-h-60 overflow-y-auto"
                        dangerouslySetInnerHTML={{ __html: highlightedEvidence }}
                      />
                    </div>

                    {/* Anchors */}
                    <div className="mb-6">
                      <h3 className="text-md font-medium text-gray-900 mb-2">Anchors</h3>
                      <ul className="list-disc pl-5">
                        {finding.anchors.map((anchor, index) => (
                          <li key={index} className="mb-1">
                            {anchor.text} (Page: {anchor.page}, Offset: {anchor.offset})
                          </li>
                        ))}
                      </ul>
                    </div>

                    {/* Verdict and Rationale */}
                    <div className="mb-6">
                      <h3 className="text-md font-medium text-gray-900 mb-2">Verdict</h3>
                      <p className="mb-2">
                        <span className="font-medium">Verdict:</span> {finding.verdict}
                      </p>
                      <p>
                        <span className="font-medium">Rationale:</span> {finding.rationale}
                      </p>
                    </div>
                  </div>
                </div>
                <div className="flex flex-shrink-0 justify-end px-4 py-4">
                  <button
                    type="button"
                    className="mr-2 inline-flex justify-center rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
                    onClick={handleCopyToClipboard}
                  >
                    Copy to Clipboard
                  </button>
                  <button
                    type="button"
                    className={`mr-2 inline-flex justify-center rounded-md px-3 py-2 text-sm font-semibold text-white shadow-sm ${
                      isReviewed ? 'bg-green-600 hover:bg-green-500' : 'bg-gray-600 hover:bg-gray-500'
                    }`}
                    onClick={handleMarkReviewed}
                  >
                    {isReviewed ? 'Reviewed' : 'Mark as Reviewed'}
                  </button>
                  <button
                    type="button"
                    className="inline-flex justify-center rounded-md bg-blue-600 px-3 py-2 text-sm font-semibold text-white shadow-sm hover:bg-blue-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-blue-600"
                    onClick={onClose}
                  >
                    Close
                  </button>
                </div>
              </div>
            </DialogPanel>
          </div>
        </div>
      </div>
    </Dialog>
  );
}