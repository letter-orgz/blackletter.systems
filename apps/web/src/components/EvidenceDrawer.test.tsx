import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import EvidenceDrawer from './EvidenceDrawer';
import type { Finding } from '@/lib/types';

beforeAll(() => {
  (global as any).ResizeObserver = class {
    observe() {}
    unobserve() {}
    disconnect() {}
  };
});

const baseFinding: Finding = {
  id: '1',
  title: 'art28_data_categories',
  verdict: 'ok',
  evidence: 'The Processor shall process personal data...',
  rationale: 'Because <strong>evidence</strong> shows compliance.',
  anchors: [{ text: 'personal data', page: 1, offset: 24 }],
  citations: [{ page: 3, text: 'Sample citation', start: 10, end: 20 }],
};

describe('EvidenceDrawer', () => {
  it('renders with anchors', () => {
    const { container } = render(
      <EvidenceDrawer isOpen onClose={() => {}} finding={baseFinding} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with citations', () => {
    const finding: Finding = { ...baseFinding, anchors: [], citations: baseFinding.citations };
    const { container } = render(
      <EvidenceDrawer isOpen onClose={() => {}} finding={finding} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with empty rationale', () => {
    const finding: Finding = { ...baseFinding, rationale: undefined };
    const { container } = render(
      <EvidenceDrawer isOpen onClose={() => {}} finding={finding} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders loading state', () => {
    const { container } = render(
      <EvidenceDrawer isOpen onClose={() => {}} finding={null} />
    );
    expect(container).toMatchSnapshot();
  });
});

