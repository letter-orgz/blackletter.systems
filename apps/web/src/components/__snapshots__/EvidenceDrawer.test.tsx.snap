// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EvidenceDrawer > renders loading state 1`] = `<div />`;

exports[`EvidenceDrawer > renders with anchors 1`] = `
<div
  aria-hidden="true"
>
  <div
    hidden=""
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`EvidenceDrawer > renders with citations 1`] = `
<div
  aria-hidden="true"
>
  <div
    hidden=""
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`EvidenceDrawer > renders with empty rationale 1`] = `
<div
  aria-hidden="true"
>
  <div
    hidden=""
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;
